### PR TITLE
Use buffered channel to unblock lazySet's goroutine

### LINKED ISSUE #1832: Lazy querier can have tens of thousands of stuck goroutines
I suspect when something goes wrong the querier stops reading responses so all the remaining ones are abandoned on an error send.

Profile shows them stuck here:

https://github.com/cortexproject/cortex/blob/6f74b9c4bcffcce6e4149a354fc4cbf2dc390a2f/pkg/querier/lazy_querier.go#L28

image

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR makes `lazyQuerier` cancellable via context.

**Which issue(s) this PR fixes**:
Fixes #1832

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/lazy_querier.go [hunk 1]
```diff
@@ -75,28 +85,43 @@ func (e errSeriesSet) Err() error {
...
+		return nil
 	}
 	return s.next.At()
 }
 
 func (s lazySeriesSet) Err() error {
-	if s.next == nil {
-		s.next = <-s.future
+	if !s.waitForSeriesSet() {
+		return nil
```
Comment (pracucci) [inline pkg/querier/lazy_querier.go]: Is it correct return `nil` here? Shouldn't we return the context error?

File: pkg/querier/lazy_querier.go [hunk 2]
```diff
@@ -11,26 +11,36 @@ import (
...
 		} else {
-			future <- set
+			res = set
+		}
+		select {
+		case future <- res:
+			// ok
+		case <-l.ctx.Done():
+			// stop
 		}
```
Comment (pracucci) [inline pkg/querier/lazy_querier.go]: Could you explain why we need this and what it does? Isn't `l.next.Select()` blocking?

Comment (pstibrany) [reply pkg/querier/lazy_querier.go]: You mean entire lazy set thing? There is comment about it in commit 76cc27152feeed512434a8a8506311f0e2099bbd, where it was added.

Now the problem with cancellation is that if one such querier returns error, another one will never get its `Next` method called, so this goroutine trying to send result would not be able to do anything. Now thinking about it, much simpler solution would be to make channel with buffer size of 1. I will change this PR to do that instead. Much simpler change, and fixes the same problem.

**Comment by pstibrany** [comment]
I've dropped previous version of the fix in favor of simply adding a buffer of size one to the channel.

**Comment by pstibrany** [comment]
> LGTM. Is it worth a mention in the `CHANGELOG`?

Doesn't really feel like it to me.

**Pull request merged**